junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TimeoutInvocation.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2015-2019 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * https://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.jupiter.engine.extension;
+
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Supplier;
+
+import org.junit.jupiter.api.extension.InvocationInterceptor.Invocation;
+
+class TimeoutInvocation<T> implements Invocation<T> {
+
+	private final Invocation<T> delegate;
+	private final TimeoutDuration timeout;
+	private final ScheduledExecutorService executor;
+	private final Supplier<String> descriptionSupplier;
+
+	TimeoutInvocation(Invocation<T> delegate, TimeoutDuration timeout, ScheduledExecutorService executor,
+			Supplier<String> descriptionSupplier) {
+		this.delegate = delegate;
+		this.timeout = timeout;
+		this.executor = executor;
+		this.descriptionSupplier = descriptionSupplier;
+	}
+
+	@Override
+	public T proceed() throws Throwable {
+		InterruptTask interruptTask = new InterruptTask(Thread.currentThread());
+		ScheduledFuture<?> future = executor.schedule(interruptTask, timeout.getValue(), timeout.getUnit());
+		Throwable failure = null;
+		T result = null;
+		try {
+			result = delegate.proceed();
+		}
+		catch (Throwable t) {
+			failure = t;
```

Comment:
I wonder what should happen here if the throwable is blacklisted. IMO in that situation it should be thrown as-is so that if, for example, it's an OOME, then the JVM can crash relatively gracefully.

@marcphilipp Thoughts?

Reply:
Agreed, we shouldn't wrap blacklisted throwables like OOMEs. 👍